tests/python/relax/test_training_optimizer.py

```diff
@@ -0,0 +1,586 @@
+    def sgd_expected(
+        params: R.Tuple(R.Tensor((3, 3), "float32"), R.Tensor((3,), "float32")),
+        gradients: R.Tuple(R.Tensor((3, 3), "float32"), R.Tensor((3,), "float32")),
+        optim_states: R.Tuple(R.Tensor((), "int64")),
+    ) -> R.Tuple(
+        R.Tuple(R.Tensor((3, 3), "float32"), R.Tensor((3,), "float32")),
+        R.Tuple(R.Tensor((), "int64")),
+    ):
+        # block 0
+        with R.dataflow():
+            lv: R.Tensor((3, 3), "float32") = params[0]
+            lv1: R.Tensor((3,), "float32") = params[1]
+            lv2: R.Tensor((3, 3), "float32") = gradients[0]
+            lv3: R.Tensor((3,), "float32") = gradients[1]
+            lv4: R.Tensor((), "int64") = optim_states[0]
+            lv5: R.Tensor((), "int64") = R.add(lv4, R.const(1, "int64"))
+            lv6: R.Tensor((3, 3), "float32") = R.multiply(R.const(0.01, "float32"), lv2)
+            lv7: R.Tensor((3, 3), "float32") = R.subtract(lv, lv6)
+            lv8: R.Tensor((3,), "float32") = R.multiply(R.const(0.01, "float32"), lv3)
+            lv9: R.Tensor((3,), "float32") = R.subtract(lv1, lv8)
+            gv: R.Tuple(R.Tensor((3, 3), "float32"), R.Tensor((3,), "float32")) = (lv7, lv9)
+            gv1: R.Tuple(R.Tensor((), "int64")) = (lv5,)
+            R.output(gv, gv1)
+        return (gv, gv1)
+
+    assert_structural_equal(sgd, sgd_expected)
+
+
+def test_sgd_complex():
+    x = relax.Var("x", R.Tensor((3, 3), "float32"))
+    y = relax.Var("y", R.Tensor((3,), "float32"))
+    sgd = SGD([x, y], 0.01, 0.02).get_function()
+
+    @R.function
+    def sgd_expected(
+        params: R.Tuple(R.Tensor((3, 3), "float32"), R.Tensor((3,), "float32")),
+        gradients: R.Tuple(R.Tensor((3, 3), "float32"), R.Tensor((3,), "float32")),
+        optim_states: R.Tuple(R.Tensor((), "int64")),
+    ) -> R.Tuple(
+        R.Tuple(R.Tensor((3, 3), "float32"), R.Tensor((3,), "float32")),
+        R.Tuple(R.Tensor((), "int64")),
+    ):
+        # block 0
+        with R.dataflow():
+            lv: R.Tensor((3, 3), "float32") = params[0]
+            lv1: R.Tensor((3,), "float32") = params[1]
+            lv2: R.Tensor((3, 3), "float32") = gradients[0]
+            lv3: R.Tensor((3,), "float32") = gradients[1]
+            lv4: R.Tensor((), "int64") = optim_states[0]
+            lv5: R.Tensor((), "int64") = R.add(lv4, R.const(1, "int64"))
+            lv6: R.Tensor((3, 3), "float32") = R.multiply(R.const(0.02, "float32"), lv)
+            lv7: R.Tensor((3, 3), "float32") = R.add(lv6, lv2)
+            lv8: R.Tensor((3, 3), "float32") = R.multiply(R.const(0.01, "float32"), lv7)
+            lv9: R.Tensor((3, 3), "float32") = R.subtract(lv, lv8)
+            lv10: R.Tensor((3,), "float32") = R.multiply(R.const(0.02, "float32"), lv1)
+            lv11: R.Tensor((3,), "float32") = R.add(lv10, lv3)
+            lv12: R.Tensor((3,), "float32") = R.multiply(R.const(0.01, "float32"), lv11)
+            lv13: R.Tensor((3,), "float32") = R.subtract(lv1, lv12)
```

Comment:
We should consider then naming in follow-up PRs, as `lv_x` is too hard to understand the logic